editor/gui/scene_tree_editor.cpp

```diff
@@ -226,46 +226,180 @@ void SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 			part_of_subscene = true;
-			//allow
+			// allow.
 		} else {
+			if (I) {
+				// stale node, remove recursively.
+				_node_cache_remove(p_node, true);
+			}
 			return;
 		}
 	} else {
 		part_of_subscene = p_node != get_scene_node() && get_scene_node()->get_scene_inherited_state().is_valid() && get_scene_node()->get_scene_inherited_state()->find_node_by_path(get_scene_node()->get_path_to(p_node)) >= 0;
 	}
 
-	TreeItem *item = tree->create_item(p_parent);
+	TreeItem *item;
+
+	bool is_new = false;
+
+	if (I) {
+		item = I->value.item;
+	} else {
+		// We don't set the node_index cache here. As undo/redo might've just recreated this.
+		item = tree->create_item(p_parent);
+		CachedNode cached_node;
+		cached_node.item = item;
+		I = node_cache.insert(p_node, cached_node);
+		is_new = true;
+
+		// If we have children we might have just been added because of a change type operation.
+		// We need to re-parent all of our children.
+		int cc = p_node->get_child_count(false);
+		for (int i = 0; i < cc; i++) {
+			Node *c = p_node->get_child(i, false);
+			HashMap<Node *, CachedNode>::Iterator IC = node_cache.find(c);
+
+			if (IC) {
+				if (IC->value.item->get_parent() != item) {
+					TreeItem *p = IC->value.item->get_parent();
+					if (p) {
+						p->remove_child(IC->value.item);
+					}
+					item->add_child(IC->value.item);
+				}
+			}
+		}
+	}
+
+	bool dirty = force_update || I->value.dirty;
 
-	item->set_text(0, p_node->get_name());
-	item->set_text_overrun_behavior(0, TextServer::OVERRUN_NO_TRIMMING);
+	if (p_parent && dirty) {
+		bool reparent = false;
+		int current_node_index = p_node->get_index(false);
+
+		if (item->get_parent() != p_parent) {
+			TreeItem *p = item->get_parent();
+			if (p) {
+				item->get_parent()->remove_child(item);
+			}
+			p_parent->add_child(item);
+			reparent = true;
+		}
+
+		if (I->value.index != current_node_index || reparent) {
+			int child_index = CLAMP(current_node_index, 0, p_parent->get_child_count() - 1);
+
+			TreeItem *new_neigbor = p_parent->get_child(child_index);
+			if (new_neigbor != item) {
+				if (child_index == p_parent->get_child_count() - 1) {
+					item->move_after(new_neigbor);
+				} else {
+					item->move_before(new_neigbor);
+				}
+			}
+		}
+		I->value.index = current_node_index;
+	} else {
+		I->value.index = 0;
+	}
+
+	if (dirty) {
+		_update_node(p_node, item, part_of_subscene);
+		I->value.dirty = false;
+
+		// Force update all our children if we are new, the number of children changed, or if we were forced to update.
+		//bool children_changed = p_node->get_child_count(false) != item->get_child_count();
+		bool force_update_children = force_update || is_new;
+		// Update all our children.
+		for (int i = 0; i < p_node->get_child_count(); i++) {
+			_update_node_subtree(p_node->get_child(i), item, force_update_children);
+		}
+
+	} else {
+		// A parent might have moved/renamed.
+		item->set_metadata(0, p_node->get_path());
```

Comment:
That could be done, would save having to update the node path on moves too. But it would make the PR even larger. I had a version that did this, but especially now it doesn't make much difference. I am agnostic on whether it should be in this PR or a separate PR. Just let me know.